### ISSUE TITLE
Why need to reset vote during pre-vote 

### ISSUE THREAD

**Issue by stone-zhang-22** [canonical/raft#196]
In the pre-vote process, the vote is set to 0. At the same time, there is a test case(name is preVoteWithcandidateCrash) for this scenario. I don’t understand why need to reset vote. Can someone answer it, please?

**Comment by freeekanayaka** [comment]
I'm not sure to understand this question. I assume you are familiar with the pre-vote algorithm as described in the Raft dissertation. The pre-vote is a "transient" state, so it must be reset if the election doesn't actual get to the vote phase. Can you please point the exact code which you don't understand?

**Comment by stone-zhang-22** [comment]
int electionStart(struct raft *r)
{
    raft_term term;
    size_t n_voters;
    size_t voting_index;
    size_t i;
    int rv;
    assert(r->state == RAFT_CANDIDATE);

    n_voters = configurationVoterCount(&r->configuration);
    voting_index = configurationIndexOfVoter(&r->configuration, r->id);

    /* This function should not be invoked if we are not a voting server, hence
     * voting_index must be lower than the number of servers in the
     * configuration (meaning that we are a voting server). */
    assert(voting_index < r->configuration.n);

    /* Sanity check that configurationVoterCount and configurationIndexOfVoter
     * have returned something that makes sense. */
    assert(n_voters <= r->configuration.n);
    assert(voting_index < n_voters);

    /* During pre-vote we don't actually increment term or persist vote, however
     * we reset any vote that we previously granted since we have timed out and
     * that vote is no longer valid. */
    if (r->candidate_state.in_pre_vote) {
        **_/* Reset vote */
        rv = r->io->set_vote(r->io, 0);
        if (rv != 0) {
            goto err;
        }
        /* Update our cache too. */
        r->voted_for = 0;_**
    } else {
        /* Increment current term */
        term = r->current_term + 1;
        rv = r->io->set_term(r->io, term);
        if (rv != 0) {
            goto err;
        }

        /* Vote for self */
        rv = r->io->set_vote(r->io, r->id);
        if (rv != 0) {
            goto err;
        }

        /* Update our cache too. */
        r->current_term = term;
        r->voted_for = r->id;
    }

    /* Reset election timer. */
    electionResetTimer(r);

    assert(r->candidate_state.votes != NULL);

    /* Initialize the votes array and send vote requests. */
    for (i = 0; i < n_voters; i++) {
        if (i == voting_index) {
            r->candidate_state.votes[i] = true; /* We vote for ourselves */
        } else {
            r->candidate_state.votes[i] = false;
        }
    }
    for (i = 0; i < r->configuration.n; i++) {
        const struct raft_server *server = &r->configuration.servers[i];
        if (server->id == r->id || server->role != RAFT_VOTER) {
            continue;
        }
        rv = electionSend(r, server);
        if (rv != 0) {
            /* This is not a critical failure, let's just log it. */
            tracef("failed to send vote request to server %llu: %s", server->id,
                   raft_strerror(rv));
        }
    }

    return 0;

err:
    assert(rv != 0);
    return rv;
}       
See the italic part of the function above. I think that there is a risk when the node crash and restart. The the node will forget who it has voted for if  we reset vote as above. This may lead to the possibility of a node casting multiple votes to different nodes in a term.

**Comment by freeekanayaka** [comment]
The reset will happen after an election timeout, therefore that vote is no longer valid and **must** be reset, so the risk is actually the opposite: if you don't reset it, it would become a "stale" vote, see [this commit](https://github.com/canonical/raft/commit/9c70d96ce201e37eed2868e60e9074fee0c3d524).

Note that if the node crashes after that reset and then votes for someone else after restarting, that's still ok, because this is a pre-vote phase, not the actual final election, in which double voting is still prevented.

**Comment by as58545652** [comment]
I think the pre-vote phase has an issue:
if a server becomes a candidate, no matter how big its term is, other servers which have voted for someone else will not vote for the candidate in the pre-vote  phase until they also become candidates and clear their voted_for. This means that if there is a leader crashes, only after a majority of servers have become candidates and cleared their voted_for, do they vote for other servers in the pre-vote phase. In other words, if the first leader crashes, any successful election will take longer time than the first one.

**Comment by as58545652** [comment]
I think the pre-vote phase has an issue:
if a server becomes a candidate, no matter how big its term is, other servers which have voted for someone else will not vote for the candidate in the pre-vote phase until they also become candidates and clear their voted_for. This means that if a leader crashes, only after a majority of servers have become candidates and cleared their voted_for, do they vote for other servers in the pre-vote phase. In other words, if the first leader crashes, any successful election will take longer time than the first one.

**Comment by freeekanayaka** [comment]
> I think the pre-vote phase has an issue:
> if a server becomes a candidate, no matter how big its term is, other servers which have voted for someone else will not vote for the candidate in the pre-vote phase until they also become candidates and clear their voted_for. This means that if a leader crashes, only after a majority of servers have become candidates and cleared their voted_for, do they vote for other servers in the pre-vote phase. In other words, if the first leader crashes, any successful election will take longer time than the first one.

Yes, that's the intended behavior. But it's unrelated to pre-vote, it would be the same even with pre-vote switched of. The reason is that if a follower knows about a leader and receives regular heartbeats from it, it will refuse to grant a vote, no matter how large is the term the candidate is presenting it. This is modification of the "plain" election algorithm which is described in the Raft dissertation and its purpose is to avoid a removed or partially disconnected server from disrupting an otherwise healthy cluster (see the 4.2.3  "Disruptive servers" section in the Raft disseration).

**Comment by freeekanayaka** [comment]
Note the subtle detail that the above mechanism doesn't mean that all followers need to become candidates before granting the vote, because they will start granting the vote as soon as the **minumum** election timeout has elapsed without hearing back from the leader, while they will convert to candidates only when the **randomized** election timeout has elapsed (the randomized is between the minimum and twice the minimum). This behavior should be the same no matter whether pre-vote is on or off.

**Comment by freeekanayaka** [comment]
@as58545652 hm, actually after looking at the code closer, you might have point. I thought the vote would be reset by a follower also after the minimum election timeout has elapsed (and the follower basically transitions to a state where it has no leader, but it does not yet convert to candidate), but that does not appear to be the case.

@MathieuBordere it is probably worth giving a deeper look at this issue, to see if it actually holds. What I described above is roughly the behavior that should be in place, but perhaps it's not. Maybe the `tickFollower` function should be enhanced to also check if the minimum election timeout has elapsed and clear both `r->follower_state.current_leader.id` and `r->voted_for` if so (and call `r->io->set_vote(r->io, 0)` too).

I'm just hand-waving here, I really didn't look too much into this, but as I said @as58545652 might be onto something.

**Comment by MathieuBordere** [comment]
Okay, I'll label it as a bug for now.

**Comment by as58545652** [comment]
If you allow followers to grant a vote after the minumum election timeout has elapsed without hearing back from the leader, it might raise another issue. For example, if there are 5 servers in a cluster, S4 and S5 are partitioned and S1 becomes leader after S2 and S3 granted their votes for it. The network is unstable. S5 just grants its vote for S4, and then loses connection with S4 and connects back with S1. Meanwhile, S2 loses connection with S1 for a minimum election timemout, and then grants its vote to S4 just after S5 has connected with S1. Because S4 receives votes from S2 and S5, so it converts to the leader. And because S1 can hear from S3 and S5, it won't step down. At this point, the  network of the cluster resumes and each server can communicate with one another. So, I think there is a possibility that a cluster might have more than one leader with the same term at some point.

**Comment by as58545652** [comment]
If you allow followers to grant a vote after the minumum election timeout has elapsed without hearing back from the leader, it might raise another issue. For example, if there are 5 servers in a cluster, S4 and S5 are partitioned and S1 becomes leader after S2 and S3 granted their votes to it. The network is unstable. S5 just grants its vote to S4, and then loses connection with S4 and connects back with S1. Meanwhile, S2 loses connection with S1 for a minimum election timemout, and then grants its vote to S4 just after S5 has connected with S1. Because S4 receives votes from S2 and S5, it converts to the leader. And because S1 can hear from S3 and S5, it won't step down. At this point, the  network of the cluster resumes and each server can communicate with one another. So, I think there is a possibility that a cluster might have more than one leader with the same term at some point.

**Comment by freeekanayaka** [comment]
@as58545652 you're right, what I described above would lead to a violation of the voting rules. So I amend that: after the minimum election timeout the follower should **not** reset the vote (by zeroing `r->voted_for` and calling `r->io->set_vote(r->io, 0)`), it should just set `r->follower_state.current_leader.id` to `0`. We also need another change: if pre-vote is enabled the follower should grant its vote if the candidate has a higher term, while I believe in the current code the vote would be refused because during pre-vote the follower does not bump its term to match the candidate one (that part is correct, since pre-vote is only advisory and should not cause any hard state changes).

**Comment by as58545652** [comment]
@freeekanayaka do you think the example I gave could also be applied to the scenario @stone-zhang-22 mentioned? It is possible that a follower converts to a candidate just after a minimum election timemout has elapsed. If you reset its vote, do you think it might grant its vote to others rather than the current leader before it votes for itself? I do not know if the Raft dissertation does mention that a server needs to reset its vote (though I know a candidate needs to vote for itself after it increments its term, and 4.2.3 also does not mention that), so I think there might be a risk.

**Comment by freeekanayaka** [comment]
> @freeekanayaka do you think the example I gave could also be applied to the Pre-Vote phase? It is possible that a follower converts to a candidate just after a minimum election timemout has elapsed. If you reset its vote, is it possible that the server grants its vote to others before it votes for itself? I do not know if the Raft dissertation does mention that a server needs to reset its vote (though I know a candidate needs to vote for itself after it increments its term (see the 3.4 Leader election, the 2nd paragraph), and 4.2.3 does not mention that either), so I think there might be a risk.

Good point. I believe it's possible that the problem you describe applies to the pre-vote phase, so the current code might be buggy. This requires a bit of thought, and maybe also checking what other raft implementations are doing during pre-vote.

**Comment by freeekanayaka** [comment]
So yeah after thinking about this a bit more, I believe the following changes should fix the situation:

1) In `electionStart()` remove the code that resets the vote if pre-vote is enabled, that leads to a violation of the Raft algorithm (because it makes it possible to double vote during the same term)

2) Set `r->follower_state.current_leader.id` to `0` after a minimum election timeout has elapsed without hearing from the leader.

3) Change `electionVote()` to still grant the vote if `args->pre_vote` is true and `args->term` is strictly larger than `r->current_term`, even if the follower has already a vote (that vote was for `r->current_term`, so granting the vote now it just means that the follower would be willing to vote if it sees a higher term)

Note that granting the vote in 3) does not mean overwriting the existing vote yet, because no state change gets actually executed/persisted during pre-vote.

**Comment by as58545652** [comment]
@freeekanayaka another issue with the pre-vote is that the current code does not distinguish results of pre-vote requests from those of normal election vote requests. I believe it is possible that a pre-vote result might be regarded as a normal one, which would possibly lead to double voting. What's your opinion?

**Comment by freeekanayaka** [comment]
> @freeekanayaka another issue with the pre-vote is that the current code does not distinguish results of pre-vote requests from those of normal election vote requests. I believe it is possible that a pre-vote result might be regarded as a normal one, which would possibly lead to double voting. What's your opinion?

I think you might be right indeed. I believe what you describe is possible with the current code, and should be investigated and possibly fixed too.

**Comment by MathieuBordere** [comment]
I'm looking into the pre-vote issues atm, and can reproduce some of the problems mentioned in this thread. The pre-vote algorithm is practically not described in the paper and because of that, it's bug prone and it makes the code more complex in one of the most critical parts of the algorithm, the voting part. This might sound drastic, but I'm thinking about removing the option altogether as it becomes increasingly difficult to reason about.

**Comment by freeekanayaka** [comment]
I don't recall exactly why atm,but I believe introducing pre-voting was necessary to solve a situation that was hit in real world and was critical. I'm afraid that I can't provide more details, I'll do if something pops up in my mind, however I'd recommend fixing the situation instead. If you look at other Raft implementations (like etcd/raft) you might get more insight about what they do to implement pre-voting correctly (and also if you dig in their issues perhaps find out similar reasons for which they had to introduce it).

**Comment by MathieuBordere** [comment]
are you referring to this https://github.com/canonical/raft/issues/65?

**Comment by MathieuBordere** [comment]
> > are you referring to this #65?
> 
> That issue was filed as feature request and I did not plan to work on it because at that time it just seemed an optimization for possibly unstable networks (e.g. WANs). I then implemented it after seeing some real world issue unrelated to unstable networks.
> 
> Now that I think of it might be when automatic role management was introduced: if you have a non-leader voter node that gets demoted to spare by the leader I don't recall exactly when the node gets to learn that it was demoted, but during that time the leader will stop sending it heartbeats and so the demoted node might start an election. I have the impression that this issue showed up in the LXD clustering test cases of the test suite (probably both on Jenkins and locally). You might want to try to reason around this, and possibly make some experiments disabling pre-vote and seeing how the LXD test suite behave or some manual testing: for example when a non-leader voter LXD node (or an `app.App` based node like the dqlite-demo) gets shutdown it first ask the leader to transfer its voting rights to a spare or stand-by node (assuming one is available) then the process ends. When you restart the node, it will probably disrupt the cluster if there's no pre-vote enabled, since it still thinks it's a voter.
> 
> I'm note entirely sure that the above is accurate, but I remember something along these lines, so it's worth checking. Turning off pre-vote and running many times both the LXD clustering tests and the go-dqlite `test/roles.sh` test might shed some light too.

Thanks for the info, I'll experiment with it & think about situations where the `disrupt_leader` flag would not suffice.

**Comment by freeekanayaka** [comment]
Not sure what you mean, but just for clarity, `disrupt_leader` has its own purpose, namely to transfer leadership when a voter gets shutdown. That's different than pre-vote, which prevents a demoted non-leader voter to disrupt the cluster (especially upon shutdown/restart).

**Comment by MathieuBordere** [comment]
Indeed, the `disrupt_leader` flag helps prevent disturbing the cluster when sending RequestVote RPC's, but a server can still disrupt the cluster by answering to AppendEntries RPC's with a higher term, that's not covered by the `disrupt_leader` flag, the higher term would be prevented by the pre-vote phase, I hadn't thought of that yet.

**Comment by MathieuBordere** [comment]
> @freeekanayaka another issue with the pre-vote is that the current code does not distinguish results of pre-vote requests from those of normal election vote requests. I believe it is possible that a pre-vote result might be regarded as a normal one, which would possibly lead to double voting. What's your opinion?
@as58545652 @freeekanayaka 
I think this is safe because if the response of a pre-vote request is received while the candidate has actually already moved beyond the pre-vote phase, then the term of the responder will be smaller than the term of the candidate that has increased its term as a result of moving out of pre-vote. The candidate will reject the response, and will not count a pre-vote vote as an actual vote.

**Comment by MathieuBordere** [comment]
> @freeekanayaka another issue with the pre-vote is that the current code does not distinguish results of pre-vote requests from those of normal election vote requests. I believe it is possible that a pre-vote result might be regarded as a normal one, which would possibly lead to double voting. What's your opinion?

I deleted my previous comment by accident, it was something like:
I think this is actually safe, a Candidate cannot count a pre-vote result as a regular vote result, because the term of the pre-vote result will be smaller than the term of the candidate that has now incremented its term as a result of moving out of the pre-vote phase.

**Comment by freeekanayaka** [comment]
Quick comment: @MathieuBordere did you make an experiment (or unit test) to check that it's really the case? My memory might be failing badly, but I would expect the term of the pre-vote and vote to be the same, because the pre-vote "artificially" bumps the term (since that's the perspective term of the election). IIRC the issue mentioned by @as58545652 is a valid one. I didn't check the code again, but that was my feeling when I read it back then to check the situation.

**Comment by MathieuBordere** [comment]
> Quick comment: @MathieuBordere did you make an experiment (or unit test) to check that it's really the case? My memory might be failing badly, but I would expect the term of the pre-vote and vote to be the same, because the pre-vote "artificially" bumps the term (since that's the perspective term of the election). IIRC the issue mentioned by @as58545652 is a valid one. I didn't check the code again, but that was my feeling when I read it back then to check the situation.

@freeekanayaka I think you're right, it looks buggy.

There are 3 servers, C, the candidate and V1 and V2, 2 voters.

When starting an election, C sends a `RequestVote RPC` with Term = Term<sub>C</sub> + 1 https://github.com/canonical/raft/blob/ad00295ee0557b9ec974d035a4ef880e145c0e40/src/election.c#L70 and with `pre-vote` set to true towards V1 and V2.

V1 answers first.
Upon reception of the `RequestVote RPC`, V1 will grant it's vote if V1 doesn't have a leader and if C's log is sufficiently up-to-date and if Term<sub>V1</sub> <= Term<sub>C</sub> + 1. V1 will send a `RequestVoteResult RPC` containing `vote_granted = true` and term = Term<sub>V1</sub>. There are 4 cases based on Term<sub>V1</sub>

#### Pre-vote
1. Term<sub>V1</sub> < Term<sub>C</sub> : C will reject the RPC
2. Term<sub>V1</sub> = Term<sub>C</sub> : C will accept the RPC, count it towards the quorum, and move out of pre-vote phase.
3. Term<sub>V1</sub> = Term<sub>C</sub> + 1 : C will accept the RPC, count it towards the quorum, and move out of pre-vote phase.
https://github.com/canonical/raft/blob/ad00295ee0557b9ec974d035a4ef880e145c0e40/src/recv_request_vote_result.c#L65
4. Term<sub>V1</sub> > Term<sub>C</sub> + 1 : C will set its term to Term<sub>V1</sub> and convert to follower https://github.com/canonical/raft/blob/ad00295ee0557b9ec974d035a4ef880e145c0e40/src/recv_request_vote_result.c#L67

C Moves out of pre-vote and bumps its term to Term<sub>Cnew</sub> = Term<sub>C</sub> + 1, now is the moment we can potentially count pre-vote `RequestVoteResult RPC` responses as regular votes. Say we now receive a `RequestVoteResult RPC` from V2 from our earlier pre-vote round, V2 was in one of the 4 situations described above to grant its vote, except now, when handling the response, C has bumped its term by 1.

#### Real-Vote
1. Term<sub>V2</sub> < Term<sub>C</sub> < Term<sub>Cnew</sub> : C will reject the RPC.
2. Term<sub>V2</sub> = Term<sub>C</sub> < Term<sub>Cnew</sub> : C will reject the RPC.
3. Term<sub>V2</sub> = Term<sub>C</sub> + 1 = Term<sub>Cnew</sub> : C will accept the RPC, count it towards the quorum of a real vote. <- **BUG**
4. Term<sub>V2</sub> > Term<sub>C</sub> + 1 = Term<sub>Cnew</sub> : C will accept the RPC, count it towards the quorum of a real vote. <- **BUG** (not sure if this is actually possible)

To fix this, I think in situation 3 above, in the pre-vote round, C should revert to follower when it sees Term<sub>V1</sub> = Term<sub>C</sub> + 1 and set its term to Term<sub>V1</sub> according to `Rules for Servers` Figure 2 in the raft paper
>If RPC request or response contains term T > currentTerm:
>set currentTerm = T, convert to follower (§5.1)

and start a new election when it times out again.

Does this make sense you think?

**Comment by freeekanayaka** [comment]
I'm not sure I understand. Let's say you make C revert to follower when it sees Term<sub>V1</sub> = Term<sub>C</sub> + 1, how does that solve the problem in the case that (say) Term<sub>V1</sub> = Term<sub>C</sub>? A late response from V2 could still arrive after C has exited the pre-vote phase upon receiving V1's vote.

What do other implementations do (etcd/raft in particular) ? Are there more details about this in the Raft paper? My gut feeling would be that the `RequestVoteResult RPC` should "echo" back the pre-vote flag, so C knows what kind of response is getting. But maybe there are other ways to deal with pre-vote that don't require that (I myself assumed that echoing the pre-vote flag was not necessary, when I first implemented the feature).

**Comment by MathieuBordere** [comment]
etcd/raft has a separate RPC Request/Response type for the PreVote case, it would be equivalent to echoing back pre-vote flag.

**Comment by freeekanayaka** [comment]
Right, a different message is equivalent to echoing back the flag.

Does the raft paper say anything about this? I guess the description of pre-vote is not very detailed, but perhaps a careful read might shed some more light on this.

**Comment by MathieuBordere** [comment]
> Right, a different message is equivalent to echoing back the flag.
> 
> Does the raft paper say anything about this? I guess the description of pre-vote is not very detailed, but perhaps a careful read might shed some more light on this.

Not really, it just talks about a "new phase" in the elections.

**Comment by freeekanayaka** [comment]
So yeah, perhaps adding the echo flag is the way to go. That needs to be done with backward compatibility in mind. If you look at `decodeRequestVote` in `uv_encoding.c` you find an example of how to handle adding a new field in a message:

```C
    /* Support for legacy request vote that doesn't have disrupt_leader. */
    if (buf->len == sizeofRequestVoteV1()) {
        p->disrupt_leader = false;
        p->pre_vote = false;
    } else {
        uint64_t flags = byteGet64(&cursor);
        p->disrupt_leader = (bool)(flags & 1 << 0);
        p->pre_vote = (bool)(flags & 1 << 1);
    }
```

One option would be to add a new 64 bit "flags" field to the `RequestVoteResult` RPC wire format, whose first bit could be used to encode the pre-vote flag.

**Comment by as58545652** [comment]
There is an issue with the current code:
If a candidate is partitioned just after its pre-vote has succeeded, then it will keep incrementing its term. When the candidate's network recovers, it may still disrupt the cluster.

**Comment by MathieuBordere** [comment]
> There is an issue with the current code: If a candidate is partitioned just after its pre-vote has succeeded, then it will keep incrementing its term. When the candidate's network recovers, it may still disrupt the cluster.

You're right, one possible solution would be to revert back to pre-vote candidate state when the election timeout has passed after an unsuccessful election.

**Comment by freeekanayaka** [comment]
I think there's a bit of tradeoff here. The likelihood of a partition occurring right between pre-vote and actual vote feels relatively low, while the likelihood of the election timeout kicking in because of competing candidates feels higher. If we revert the candidate to pre-vote we'll optimize for the first scenario at the expense of the second (because electing a leader will take longer).

I'm not entirely sure if it's worth "fixing", but YMMV.

**Comment by MathieuBordere** [comment]
> I think there's a bit of tradeoff here. The likelihood of a partition occurring right between pre-vote and actual vote feels relatively low, while the likelihood of the election timeout kicking in because of competing candidates feels higher. If we revert the candidate to pre-vote we'll optimize for the first scenario at the expense of the second (because electing a leader will take longer).
> 
> I'm not entirely sure if it's worth "fixing", but YMMV.

Indeed was just editing my response, we probably need to be a bit pragmatic here.
